### ISSUE TITLE
Suggestion: provide the encoding(UTF-8 with BOM) option when creating project to be Chinese friendly

### ISSUE THREAD

**Issue by CodingNinjaOctocat** [Microsoft/WindowsTemplateStudio#598]
😁

**Comment by ralarcon** [comment]
Thanks for the suggestion!

**Comment by crutkas** [comment]
@rido-min is this a VS level request or something WTS needs to set when generating files?

**Comment by mrlacey** [comment]
I had some similar issues after some of the template files were converted to ASCII while others were UTF8 and it broke some of the merge actions as the string compare couldn't find matching lines.

We don't just need to make this change for supporting non-ASCII we need to be sure we're consistent across all files.

We should also have a test to make sure that no files get changed and that new template files are the correct encoding.

Having done a quick check it looks like only about a quarter of template files are UTF8 and the rest are ASCII.

Files I think this should cover are:
- *.cs
- *.xaml
- *.json
- *.resw
- *.csproj
- *.md

Basically, anything that's text based.
Any others?

**Comment by ralarcon** [comment]
I think you have the list. IMHO this is really important as this would lead to hidden wierd behaviours... 

@crutkas I think it belong to us... we need ensure the encodings are consisten.
@mrlacey let me know if you need something from Madrid team.

**Comment by crutkas** [comment]
#587 ref

**Comment by crutkas** [comment]
if you're including csproj, what about 

- sln
- vbproj (pending)
- appxmanifest
- xml

**Comment by CodingNinjaOctocat** [comment]
I used the Chinese version VS (2017), and I tested it for some help. In my work environment, using VS to create a blank UWP project (or other non-third-party template project), the file encoding is **UTF-8-BOM**, which includes:

- *.cs
- *.xaml
- *.csproj
- *.sln
- *.appxmanifest

If i create the project with WTS,`*.cs` and `*.xaml` file encoding is **UTF-8** (will make Chinese garbled), the rest is unchanged, so the Contributor used **UTF-8**, and if WTS provides an **file encoding converter script** in the final step of the creation (before the initial commit git), it should be a better solution

**Comment by crutkas** [comment]
think i figured out how to do at least a few of the files.  This is a bit larger of a change possibly now that i'm trying to reverse engineer stuff.

**Comment by crutkas** [comment]
how are you determining if the file is utf-8 bom?

**Comment by crutkas** [comment]
I got what looks to be most but not everything.  if you just do a file->new and crank threw most appear now to be encoded but app.xaml.cs isn't.

Bed time over in Redmond so passing the ball here.
https://github.com/Microsoft/WindowsTemplateStudio/tree/Issue596_encodeWithBom_ChineseCharSupport

**Comment by crutkas** [comment]
* i used notepad++ to validate bom added, FYI

**Comment by mrlacey** [comment]
I can't find a definitive source that explains which encoding VS2017 uses but have found [acknowledgment of an issue](https://developercommunity.visualstudio.com/content/problem/21744/vs2017-rc-breaks-the-encoding-of-my-files.html) that causes it to change the encoding without warning but which should be fixed with 15.3. VS appears to base encoding on a combination of VS version, file extension, and regional settings. This would explain the inconsistency we see.

I still think we need to ensure that we default to something that works for everyone and that means making sure we can use UTF-8 with a BOM/signature.

For absolute clarity, I believe this means we need the files to start with the following bytes: **0xEF, 0xBB, & 0xBF**.

I wrote a little PowerShell script to look at everything in the templates folder. Doing this manually is too prone to errors.

I propose the following fix:
- change all existing plain text files (as identified by extensions listed above.)
- Add a test that verifies the encoding of all files.
- Also, add a script to the root of the repository which can set the encodings for any files that need changing. The test can point to this script to fix any issues it detects. This will be simpler than having to change encodings manually.

@crutkas  Does this need to be in for 1.2?

**Comment by crutkas** [comment]
My solution for the branch forces the UTF-8 with BOM but i don't think i got everything as i think some files are copied over versus written over and are already affected.  Now that i think about it however, there could be files that I'm being too aggressive on for forcing an encode.

I like your fix.

I think it would be nice to have for 1.2 but I think that is a bit more work to take on right now than planned seeing the amount of random bugs we're seeing come in.

Lets do it for 1.3 and get it into a nightly ASAP after we ship 1.2

**Comment by ralarcon** [comment]
Does this make sense to reinforce in the build process? I mean, mabey we can launch the tool in the build process ensuring everthing we release in the templates have the proper configuration...

**Comment by mrlacey** [comment]
The script could be run as part of a build step but it would probably be overkill to read _every_ template file on each build.

**Comment by crutkas** [comment]
How about do a unit test and cause a Build failure?  This way we are being proactive and ensuring the state of our files.

1. Verify all template files are UTF-8 w/ BOM 
1. Check generated projects files are UTF-8 w/ BOM for the all frameworks / project types w/ all options enabled template build unit tests.

**Comment by mrlacey** [comment]
FYI. **1190** template files currently **not** encoded correctly!

These are all just template files though.
We also need to check the output of the generator to make sure that files which are created by the wizard (such as .sln) or are read and rewritten are all encoded correctly. 

Need to change all instances of the following to use the overload which includes encoding:
- File.WriteAllText
- File.WriteAllLines

Also, need to set encoding property when using TextWriter (In Localization tool)

**Comment by ralarcon** [comment]
Should we open a different issue for the localization tool?

**Comment by mrlacey** [comment]
@ralarcon I didn't look at the localization tool code very thoroughly (aside it needs some tidying up--it's code like any other in the repo and should look the same.) but it seems to copy the original default language files so as long as the defaults get updated it should be good.
TextWriter defaults to the correct encoding so that's not an issue.

**Comment by mrlacey** [comment]
All merged in and seems good to go.
